keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,421 @@
+			}
+			klog.V(4).Infof("synchronizing PersistentVolume[%s]: claim %s found: %s", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef), getClaimStatusForLogging(claim))
+			// Check if the PV has a deletion timestamp, if so, annotate the PV to indicate that the PV can be deleted
+			// from the underlying storage.
+			if protectionutil.NeedToAddFinalizer(volume, pvutil.PvDeletionProtectionFinalizer) {
+				klog.Infof("synchronizing PersistentVolume[%s]: the volume was requested to be deleted but still" +
+					" it is still associated with the claim %s", volume.Name, claimrefToClaimKey(volume.Spec.ClaimRef))
+				_, err = ctrl.addPvDeletionProtectionFinalizer(volume)
+				if err != nil {
+					klog.V(4).Infof("synchronizing PersistentVolume[%s]: failed to add the pv deletion protection finalizer", volume.Name)
+				}
+				klog.Infof("successfully added pv deletion protection finalizer to indicate the volume is to be deleted.")
+			}
+		}
+// Remaining code below skipped for brevity...
+// Please refer kubernetes/kubernetes/pkg/controller/volume/persistentvolume/pv_controller.go for the full code.
+```
+
+When `deleteVolumeOperation` executes, the finalizer is checked, if present it implies that the PV was deleted first,
+in this scenario `DeletionTimestamp` checks can be ignored. Once the volume is deleted by the plugin the finalizer can
+be safely removed.
+
+```go
+// deleteVolumeOperation deletes a volume. This method is running in standalone
+// goroutine and already has all necessary locks.
+func (ctrl *PersistentVolumeController) deleteVolumeOperation(volume *v1.PersistentVolume) (string, error) {
+	klog.V(4).Infof("deleteVolumeOperation [%s] started", volume.Name)
+	newVolume, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Get(context.TODO(), volume.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.V(3).Infof("error reading persistent volume %q: %v", volume.Name, err)
+		return "", nil
+	}
+
+	// DeletionTimestamp would have been set if the PV was deleted first. Check if the PV deletion protection
+	// finalizer is set, if so, the DeletionTimestamp can be ignored
+	if !ctrl.hasPvDeletionProtectionFinalizer(volume, pvutil.PvDeletionProtectionFinalizer) {
+		if newVolume.GetDeletionTimestamp() != nil { ]
+			klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+			return "", nil
+		}
+	}
+    //.....
+    //.....
+    //.....
+    klog.V(4).Infof("deleteVolumeOperation [%s]: success", volume.Name)
+    // Remove the protection finalizer on a successful delete.
+    _, err = ctrl.removePvDeletionProtectionFinalizer(volume) //<=============================== Remove the finalizer once the volume has been deleted from plugin
+    if err!= nil {
+        klog.V(3).Infof("failed to remove pv deletion protection finalizer on the " +
+            "volume %q, err: %v", volume.Name, err)
+        return pluginName, err
```

Comment:
Just to make sure I understand: PV controller will retry deleting the volume after exp. backoff and since the volume plugin is idempotent, the controller will retry to remove the finalizer next time. That should work. If the volume plugins are idempotent :-). I think this is untested piece of volume plugins.